### ISSUE TITLE
18.6, 19.1 Asynchronous vs synchronous requests

### ISSUE THREAD

**Issue by ybakos** [shiffman/LearningProcessing#9]
Despite the time spent on the client / server is indeed asynchronous (a short simple request may invoke a long server-side response), the use of _asynchronous request_ is incorrectly used in 18.6 and 19.1. You also address blocking, which really means "synchronous," so this is important to keep aligned.

18.6
_a request ... is asynchronous, meaning ...
This should be _is synchronous, meaning your program waits for a response from the server before continuing_.

19.1
_The above process is known as an asynchronous request and response_
This should be _The above process is known as a synchronous request and response, because the client waits until the server responds before sending another message._

_imagine if you were to use asynchronous..._
This should be _imagine if you were to use synchronous..._

_...a different type of connection is used, a synchronous one..._
This should be _...a different type of connection is used, an asynchronous one..._

_Synchronous communication is also necessary for live..._
This should be _Asynchronous communication is also necessary for live..._

fig. 19.1 Left diagram should be _Synchronous request_.

I think you can essentially swap your use of synchronous/asynchronous without causing the need for any additional elaboration.

**Comment by shiffman** [comment]
Ah, I'm seeing the error of my ways here.   While HTTP requests are usually handled asynchronously Processing with `loadStrings()`, `loadXML()`, etc handles them synchronously.  I need to clean up this discussion and show how `thread()` allows you to handle them asynchronously.

In chapter 19 though my current edition is flawed I do mean asynchronous in my discussion of HTTP requests and it should really say something like:

Just to introduce myself, I'm Firefox, the web browser, and I have a <strong><em>request</em></strong>. I was wondering if you might be so kind as to send me your web page about vacation villas in
Tuscany? I'm going to be handling some other tasks and just alert me when it's ready and I'll come back to receive the data.

The distinction between HTTP Requests and the socket connection is not about asynchronous vs synchronous but more about persistent (a socket) vs. opening and closing for each request (HTTP).  Does this seem right @ybakos?

**Comment by ybakos** [comment]
"The distinction between HTTP Requests and the socket connection is not about asynchronous vs synchronous but more about persistent (a socket) vs. opening and closing for each request (HTTP)"

That's right, they are two separate concerns.

But I'm not quite sure where you're headed with that revision to the anthropomorphization of Firefox (and holy shit did I just finally use that word in a sentence).

To me, section 19.1 is about demonstrating the the user how synchronous communication can result in a delay, since the client must wait for the server's response before the client can continue. So according to your revision there, that spirit is lost.

In other words, I think the spirit and purpose of 19.1 is correct (tell the reader about synchronous vs asynchronous communication), it's just that the wrong terms are used in certain places.

If you make that change, starting Firefox in an asynchronous conversation, then the subsequent writing in that whole section needs changed.

**Comment by shiffman** [comment]
Yes I see what you are getting at. Thanks for this feedback it's incredibly helpful.  I'm going to try a few different ways and see what works best.  Will keep this open until it's resolved.

**Comment by shiffman** [comment]
Current text is as follows, the illustrations still need to be updated so I will leave this open.

<p>The above process is known as request and response, bi-directional communication between a server and a client where the server responds at its leisure and the client sits and waits.  Technically it's an "HTTP" (hypertext transfer protocol) request.  The connection is established temporarily in order to transfer the web page after which it is promptly closed.</p>

<p>In terms of sending information along the world wide web, this methodology is perfectly adequate. Nevertheless, imagine if you were to use this sort of communication in a multiplayer game. It would be a disaster. Having to open and close the connection every time one player sends a message to another would result in huge delays and lag between movements. For multiuser applications in <em>Processing</em> where we need near real-time communication, a different sort of connection is used known as a <em>socket</em> connection.  A socket connection is continuous and does not need to be opened and closed for each request.  In addition, Processing handles the messages asynchronously allowing the draw() loop to execute continuously and not sit and wait for each message to come in.  This sort communication is also necessary for live performance applications that need real-time interaction between elements. See <a data-type="xref" href="#ch19fig01">#ch19fig01</a>.</p>

**Comment by shiffman** [comment]
* [ ] Update Figure 01 and 02 to clarify distinction between asynchronous and synchronous and socket vs HTTP.

**Comment by ybakos** [comment]
How about:
The above process is known as request and response, a bi-directional communication between a client and a server. The client sends a request to the server, which responds at its leisure, while the client sits and waits for the response.

For the second paragraph, I see what you're communicating but this seems to misinform the user that socket connections are somehow different from other kinds of client-server connections, that socket communication is inherently asynchronous, and that HTTP connections aren't socket connections. The truth is that all network communications rely on sockets. Whether the socket is persistent or not, and whether or not the threads behind these sockets are synchronous/asynchronous are all independent matters. I realize that the Processing `Client` et al hides a lot of this.

So maybe there are two points to communicate in that paragraph regarding real-time network interaction. First, that opening and closing connections per request/response is expensive, so we want connections to remain open. Second, that we want communication to be asynchronous, so programs don't come to a halt waiting for requests/responses.

**Comment by shiffman** [comment]
Thanks, this is extremely helpful.   I've revised this again now.  I realized that the better place to formally address synchronous vs asynchronous is chapter 18 where a new discussion on the `thread()` method is discussed in the context of `loadStrings()`, `loadXML()`, etc.  Now this chapter can focus on sockets and the network library and refer back to the synchronous vs asychronous discussion.  Am going to close this and open a new issue for revising the illustrations.   Please take a look again when I send out a new PDF!
